examples/platforms/cc2652/flash.c

```diff
@@ -0,0 +1,317 @@
+{
+    uint32_t batMonCtl = HWREG(AON_BATMON_BASE + AON_BATMON_O_CTL);
+    return ((batMonCtl & AON_BATMON_CTL_CALC_EN_M)
+            == AON_BATMON_CTL_CALC_EN
+            && (batMonCtl & AON_BATMON_CTL_MEAS_EN_M)
+            == AON_BATMON_CTL_MEAS_EN);
+}
+
+/**
+ * Check if the supply voltage is high enough to support flash programming.
+ *
+ * @return If the Voltage is too low to support flash programming.
+ * @retval true  The supply voltage is too low.
+ * @retval false The supply voltage is sufficient.
+ */
+static bool isVoltageLow(void)
+{
+    bool batMonWasOff = !isBatMonOn();
+    bool ret = true;
+
+    if (batMonWasOff)
+    {
+        AONBatMonEnable();
+    }
+
+    if (AONBatMonBatteryVoltageGet() >= MIN_VDD_FLASH)
+    {
+        ret = false;
+    }
+
+    if (batMonWasOff)
+    {
+        AONBatMonDisable();
+    }
+
+    return ret;
+}
+
+/**
+ * Disable Flash data caching and instruction pre-fetching.
+ *
+ * It is necessary to disable the caching and VIMS to ensure the cache has
+ * valid data while the program is executing.
+ *
+ * @return The VIMS state before being disabled.
+ */
+static uint32_t disableFlashCache(void)
+{
+    uint32_t mode = VIMSModeGet(VIMS_BASE);
+
+    VIMSLineBufDisable(VIMS_BASE);
+
+    if (mode != VIMS_MODE_DISABLED)
+    {
+        VIMSModeSet(VIMS_BASE, VIMS_MODE_DISABLED);
+
+        while (VIMSModeGet(VIMS_BASE) != VIMS_MODE_DISABLED);
+    }
+
+    return mode;
+}
+
+/**
+ * Restore the Flash data caching and instruction pre-fetching.
+ *
+ * @param [in] mode The VIMS mode returned by @ref disableFlashCache.
+ */
+static void restoreFlashCache(uint32_t mode)
+{
+
+    if (mode != VIMS_MODE_DISABLED)
+    {
+        VIMSModeSet(VIMS_BASE, mode);
+    }
+
+    VIMSLineBufEnable(VIMS_BASE);
+}
+
+/**
+ * Translate the errors from the Flash programming FSM to OpenThread error
+ * codes.
+ *
+ * @param [in] error Return from the Flash programming function.
+ *
+ * @return The corresponding OpenThread @ref otError value.
+ */
+static otError fsmErrorToOtError(uint32_t error)
+{
+    otError ret = OT_ERROR_GENERIC;
+
+    switch (error)
+    {
+    case FAPI_STATUS_SUCCESS:
+        ret = OT_ERROR_NONE;
+        break;
+
+    case FAPI_STATUS_INCORRECT_DATABUFFER_LENGTH:
+        ret = OT_ERROR_INVALID_ARGS;
+        break;
+
+    case FAPI_STATUS_FSM_ERROR:
+        ret = OT_ERROR_FAILED;
+        break;
+
+    default:
+        break;
+    }
+
+    return ret;
+}
+
+/**
+ * Function documented in platforms/utils/flash.h
+ */
+otError utilsFlashInit(void)
+{
+    return OT_ERROR_NONE;
+}
+
+/**
+ * Function documented in platforms/utils/flash.h
+ */
+uint32_t utilsFlashGetSize(void)
+{
+    return FlashSizeGet();
+}
+
+/**
+ * Function documented in platforms/utils/flash.h
+ */
+otError utilsFlashErasePage(uint32_t aAddress)
+{
+    uint32_t mode;
+    uint32_t fsmRet;
+
+    if (isVoltageLow())
+    {
+        /* Supplied voltage was too low. */
+        return OT_ERROR_FAILED;
```

Comment:
Only one return per function, you can use one of the error handling functions here instead.